src/authentication/AuthenticationContext.tsx

```diff
@@ -0,0 +1,136 @@
+import React, {
+  createContext,
+  useState,
+  useEffect,
+  useCallback,
+  useMemo
+} from "react";
+import { useHistory } from "react-router-dom";
+import { useQueryClient } from "@tanstack/react-query";
+import { ADMIN_PANEL_RICHIESTE, DASHBOARD, LOGIN } from "../navigation/routes";
+import { adminLogoutRedirect, useLoginRedirect } from "./authentication";
+import { authenticationStore } from "./authenticationStore";
+import {
+  UserSession,
+  CurrentSession,
+  AdminSession,
+  MerchantInfo
+} from "./authenticationState";
+
+type AuthenticationContextType = {
+  userSessionByFiscalCode: Record<string, UserSession>;
+  adminSessionByName: Record<string, AdminSession>;
+  currentSession: CurrentSession;
+  currentUserFiscalCode: string | undefined;
+  currentUserSession: UserSession | undefined;
+  currentMerchantFiscalCode: string | undefined;
+  currentMerchant: MerchantInfo | undefined;
+  currentAdminSession: AdminSession | undefined;
+  changeSession(session: CurrentSession): void;
+  logout(session: CurrentSession): void;
+};
```

Comment:
Use optional (`?`) as more idiomatic than `Type | undefined` in TypeScript:
```suggestion
type AuthenticationContextType = {
  userSessionByFiscalCode: Record<string, UserSession>;
  adminSessionByName: Record<string, AdminSession>;
  currentSession: CurrentSession;
  currentUserFiscalCode?: string;
  currentUserSession?: UserSession;
  currentMerchantFiscalCode?: string;
  currentMerchant?: MerchantInfo;
  currentAdminSession?: AdminSession;
  changeSession(session: CurrentSession): void;
  logout(session: CurrentSession): void;
};
```

Reply:
adressed in c6f83f0bb27aaf3e8732ac39a56dbeaa0ed37870